astropy/utils/tests/test_data.py

```diff
@@ -1755,3 +1779,64 @@ def test_download_parallel_respects_pkgname(temp_cache, valid_urls):
                                pkgname=a)
     assert not get_cached_urls()
     assert len(get_cached_urls(pkgname=a)) == FEW
+
+
+@pytest.mark.parametrize("b", _shelve_possible_backends)
+def test_cache_with_different_shelve_backends(temp_cache, b, valid_urls):
+    """Without special handling this emits a warning for dbm.dumb."""
+    with shelve_backend(b):
+        clear_download_cache()
+
+        uc = list(islice(valid_urls, FEW))
+        for u, c in uc:
+            download_file(u, cache=True)
+
+        check_download_cache()
+
+        for u, c in uc:
+            assert get_file_contents(
+                download_file(u, cache=True, sources=[])) == c
+
+        clear_download_cache()
+
+
+@pytest.mark.parametrize("b", _shelve_possible_backends)
+def test_lock_behaviour_if_directory_disappears(temp_cache, b):
+    dldir, urlmapfn = _get_download_cache_locs()
+    try:
+        dbm = importlib.import_module(b)
+    except ImportError:
+        pytest.skip(f"module {b} not available")
+    if b == "dbm.dumb":
+        c = pytest.raises(FileNotFoundError)
+    else:
+        @contextlib.contextmanager
```

Comment:
@astrofrog added this in `master`, should you should probably use it instead of rolling your own; FYI:

https://github.com/astropy/astropy/blob/dce08d5f6236d14a3b3cee7cb715f5864df48ac9/astropy/utils/compat/context.py#L8

Reply:
Thanks!